man/cat1/window.1



WINDOW(1)                TIPSY COMMANDS                 WINDOW(1)



NAME
     window - switch to an alternative viewing window

SYNOPSIS
ALIASES
DESCRIPTION
SEE ALSO













































Katz and Quinn ReleasLeas2t.0change: 22 MARCH 1994                  1








code/ChangeLog
Wed Jan 11 13:09:36 1995  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (SRCS): added alloca.c so it will be included in the
	distribution.

Wed Dec 28 13:42:47 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* readmark.c: close file after it is read.

	* main.c: RedrawViewport(): make can_size the maximum of
	can_height and can_width.

	* canvas_resize_proc.c: Same as above.

Mon Dec  5 09:39:44 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* xray_lum_load.c, star_lum.c, star_lum_redshift.c: Changed
	argument to my_gets() to blank string.

Fri Dec  2 08:21:50 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in (READLINE_SRC): delete config.h.
	(tipsy): added dependence on ALLOCA.

	* fits.c: Added test for endianness.

Thu Dec  1 15:14:58 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* vista.c: fixed initialization of quantity[].

	* find_shape.c: Forced inertia_cm to be symmetric.  This avoid
	problems with jacobi().

Tue Nov  1 11:26:56 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* Makefile.in : fits_xray.c is now gone.  fits() is used instead.
	defs.h, xray.c, vista.c, fits.c and arguments.c are also changed.

Mon Oct 31 09:43:29 1994  Tom Quinn  <trq@hermes.astro.washington.edu>

	* main.c: RedrawColorbar(): fixed off-by-one drawing errors.

Wed Oct 12 14:27:16 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Created version 2.1.2.

	* absorb.c: New function for calculating aborbtion spectra.

Tue Sep  6 16:06:02 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline/: Replaced with readline-2.0 from prep.ai.mit.edu.

	* my_gets.c: Changed cast to match new external from readline/.

Fri Jul 29 10:25:49 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Ensure that the axes projections do not overflow.

Fri Jul 15 17:28:15 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* profile.c: Check to see if fopen() fails.

Wed Jul 13 09:46:18 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h: Decreased INTMAX to 2^29-1.  (One bit for sign, one bit
	to do INTMAX - (-INTMAX), and one bit for zoom out.

Tue Jul 12 15:55:17 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* canvas_resize_proc.c: Call plot_sub() and draw_label() for
	servers that automatically do backing store and don't generate an
	expose event.

	* main.c: ZoomIn(), ZoomOut(), Pan(), SetBox(): Ignore button
	presses in inactive windows.
	
	* main.c, window.c: Changed ResizeRequest to ConfigureNotify.

Sat Jun  4 13:12:06 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* fits.c, and fits_xray.c: Changed write() to fwrite() for
	portability.

Thu May  5 10:53:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* project.c: Increase clipping range of velocities to -2*INTMAX <
	v < 2*INTMAX.
	* defs.h: Change INTMAX to 2^30-1 to accommodate the above
	clipping.

Tue Apr 19 18:00:14 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* comm.h: NSK: added pcenter and writeboxrot commands.

Mon Apr 18 14:14:45 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Add "unzoom" button.

Fri Apr 15 16:01:56 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Fixed bug with color of marked particles.
	Deleted extra conversion to screen coordinates.
	
	* maxpos.c: Make the margins a little smaller.

Thu Mar 31 14:03:11 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Changed XDrawArc to XFillArc for vector draws.

Fri Mar  4 18:09:25 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Added Neal's changes for read_vector() and "master".

Mon Feb 28 11:51:39 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* openascii.c: Added code from Peter Teuben to uncompress files.

Thu Feb 24 13:21:50 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* Changed current_plot logical to current_project and
	current_color.

Tue Feb 22 11:24:05 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readline: Updated to readline from gdb-1.12

Thu Feb 17 11:46:07 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Implimented a software "z-buffer".  Only draw
	particles with the highest particle color.

Wed Feb 16 13:21:35 1994  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Merged plot_all and plot_vel into one routine.  A
	new routine all_color() is created.  Changes were also needed in
	the view_*() routines.

Sat Nov 13 11:35:01 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Fixed typo for increasing ndark.

Wed Nov  3 14:35:56 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* plot_all.c: Deleted skip of XClearWindow on first time through.

Thu Oct 28 10:23:50 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* hard.c: Added movie parameter.

Tue Oct 26 11:29:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* alloca.c: New file in case system doesn't provide alloca().

	* vista.c: Included <stdlib.h> for malloc() definition.

Mon Oct 25 19:16:52 1993  Tom Quinn  (trq@virgo)

	* main.c: Attempt to find a depth 8 PseudoColor visual.

Thu Oct 21 08:56:32 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* main.c: Set backgrounds of gc_color[] to black.

Tue Oct 19 10:56:36 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* defs.h, main.h, make_macro.c, read_macro.c, write_macro.c
	run_macro.c: re-vamped macro storage using linked lists.

Mon Oct 18 15:11:34 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* draworient.c: Added cast from Arg to long.

Fri Oct 15 10:04:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* readmark.c: Added a check for index < 1.

	* plot_sub.c: Added a check for binary_loaded.
	Added check for boxes_loaded[0].

Wed Oct  6 09:13:49 1993  Tom Quinn  (trq@hermes.astro.washington.edu)

	* zoom_window.c: Added zoom command.  Also changed comm.h and
	arguments.c.

Tue Oct  5 17:27:43 1993  Tom Quinn  (trq@hermes.astro.washington.edu

code/window.c
/* $Header$
 * $Log$
 * Revision 1.5  2003/06/13 17:37:37  trq
 * Replaced "include <malloc.h>" with "include <stdlib.h>".  This will allow
 * compilation on MAC OSX.  Also replaced "values.h" with "float.h".
 *
 * Revision 1.4  2001/07/20 23:10:29  trq
 * Added check for maxwins in delete_window() to prevent core dump.
 *
 * Revision 1.3  1998/02/18  22:53:06  trq
 * Always call canvas_resize_proc() to get new window height and width.
 *
 * Revision 1.2  1996/02/16  17:19:25  trq
 * Fixed bug in window for non-8 bit displays.
 * Added optional size arguments to the window command.
 *
 * Revision 1.1.1.1  1995/01/10  22:57:35  trq
 * Import to CVS
 *
 * Revision 2.7  1994/10/07  16:55:15  trq
 * Corrected use of XtGetValues().
 *
 * Revision 2.6  1994/07/13  11:47:10  trq
 * Look for ConfigureNotify event.
 *
 * Revision 2.5  94/04/20  08:56:52  trq
 * Added title variable
 * 
 * Revision 2.3  1991/07/16  16:11:15  trq
 * Added backing store code.
 *
 * Revision 2.2  1991/07/02  14:05:26  trq
 * Expanded functionallity and added delete_window().
 *
 * Revision 2.1  1991/04/12  17:21:30  trq
 * *** empty log message ***
 *
 */
#include "defs.h"
#include "fdefs.h"
#include <stdlib.h>
#include <X11/Xatom.h>
#include <X11/Intrinsic.h>
#include <X11/Xaw/Simple.h>
#include <X11/Shell.h>
#include <X11/StringDefs.h>
#include <X11/Xaw/Viewport.h>

static char vpTranslations[] = 
  "<Expose>:	RedrawViewport() \n\
   <ConfigureNotify>:	canvas_resize_proc() \n\
   None <Btn1Down>:	ZoomIn() \n\
   None <Btn2Down>:	ZoomOut() \n\
   None <Btn3Down>:	Pan() \n\
   ! Shift <Btn1Down>:	SetBox() \n\
   <MotionNotify>:	DrawBox()";

extern XtAppContext app_con;
extern Widget toplevel;
extern Widget canvas;
static Widget *shells;
static Widget *views;
static Widget *canvases;
static int *popped;
static int maxwins = 0;

void
window(job)
    char job[MAXCOMM] ;
{
    char command[MAXCOMM] ;
    int winnum;
    int i;
    String params;
    Cardinal nparams;
    int narg;
    int width, height;
    Visual *visual;
    int depth;
    
    if((narg = sscanf(job, "%s %d %d %d", command, &winnum, &width, &height)) == 2
	|| narg == 4)
      {
	XEvent event;
	Arg args[2];

	if(winnum > maxwins)
	  {
	    if(maxwins == 0) 
	      {
		shells = (Widget *)malloc(winnum*sizeof(Widget));
		views = (Widget *)malloc(winnum*sizeof(Widget));
		canvases = (Widget *)malloc(winnum*sizeof(Widget));
		popped = (int *)malloc(winnum*sizeof(int));
	      }
	    else 
	      {
		shells = (Widget *)realloc(shells, winnum*sizeof(Widget));
		views = (Widget *)realloc(views, winnum*sizeof(Widget));
		canvases = (Widget *)realloc(canvases, winnum*sizeof(Widget));
		popped = (int *)realloc(popped, winnum*sizeof(int));
	      }
	    for(i = maxwins; i < winnum; i++) 
	      {
		shells[i] = NULL;
		popped[i] = 0;
	      }
	    maxwins = winnum;
	  }

	if(winnum == 0)
	  {
	    currentview_xid = XtWindow(canvas);
	    canvas_resize_proc(canvas, &event, &params, &nparams);
	    XtSetArg( args[0], XtNbackground, &colors[0]);
	    XtGetValues(canvas, args, 1);
	  }
	else
	  {
	    if(shells[winnum-1] == NULL) 
	      {

		shells[winnum-1] = XtCreatePopupShell("tipsy_popup",
						   topLevelShellWidgetClass,
						   toplevel, NULL, 0);
		XtSetArg(args[0], XtNvisual, &visual);
		XtSetArg(args[1], XtNdepth, &depth);
		XtGetValues(toplevel, args, 2);
		XtSetArg(args[0], XtNvisual, visual);
		XtSetArg(args[1], XtNdepth, depth);
		XtSetValues(shells[winnum-1], args, 2);
		views[winnum-1] = XtCreateManagedWidget( "hardview",
					     viewportWidgetClass,
					     shells[winnum-1], NULL, 0);
		canvases[winnum-1] = XtCreateManagedWidget("hard",
						simpleWidgetClass,
						views[winnum-1], NULL, 0);
		XtOverrideTranslations(canvases[winnum-1],
				   XtParseTranslationTable(vpTranslations));
		if(winnum == 1)
		  {
		    hard_width = WidthOfScreen(XtScreen(canvas));
		    hard_height = HeightOfScreen(XtScreen(canvas));
		    XtSetArg(args[0], XtNwidth, hard_width);
		    XtSetArg(args[1], XtNheight, hard_height);
		  }
		else
		  {
		      Dimension xwidth, xheight;

		      if(narg == 2) {

			XtSetArg(args[0], XtNwidth, &xwidth);
			XtSetArg(args[1], XtNheight, &xheight);
			XtGetValues(canvas, args, 2);
		      } else {
			  xwidth = width;
			  xheight = height;
		      }
			XtSetArg(args[0], XtNwidth, xwidth);
			XtSetArg(args[1], XtNheight, xheight);
		  }
		XtSetValues( views[winnum-1], args, 2 );
		XtSetValues( canvases[winnum-1], args, 2 );
		XtSetValues( shells[winnum-1], args, 2 );
		XtRealizeWidget(shells[winnum-1]);
		XtPopup(shells[winnum-1], XtGrabNone);
		popped[winnum-1] = 1;
		if(DoesBackingStore(XtScreen(shells[winnum-1])) != NotUseful)
		  {
		    XSetWindowAttributes win_atts;
		    unsigned long mask = CWBackingStore;
		    win_atts.backing_store
		      = DoesBackingStore(XtScreen(shells[winnum-1]));
		    XChangeWindowAttributes(baseframe_dpy,
					    XtWindow(canvases[winnum-1]),
					    mask, &win_atts);
		  }
	      }
	    canvas_resize_proc(canvases[winnum-1], &event, &params,
				 &nparams);
	    if(!popped[winnum-1])
	      {
		XtPopup(shells[winnum-1], XtGrabNone);
		popped[winnum-1] = 1;
	      }
	    XClearWindow(baseframe_dpy,XtWindow(canvases[winnum-1])) ;
	    currentview_xid = XtWindow(canvases[winnum-1]);
	    XtSetArg( args[0], XtNbackground, &colors[0]);
	    XtGetValues(canvases[winnum-1], args, 1);
	  }
	while(XtAppPending(app_con))
	  {
	    XtAppNextEvent(app_con, &event);
	    XtDispatchEvent(&event);
	  }
      }
    else {
	printf("<bad arguments to window, %s>\n",title) ;
    }
}

void
delete_window(job)
    char job[MAXCOMM] ;
{
    char command[MAXCOMM] ;
    int winnum;
    String params;
    Cardinal nparams;
    
    if(sscanf(job, "%s %d", command, &winnum) == 2)
      {
	XEvent event;
      
	if(winnum == 0)
	  {
	    printf("<try quit, %s>\n",title);
	    return;
	  }
	if(winnum > maxwins || !popped[winnum-1])
	    {
		printf("<window is not open, %s>\n",title);
		return;
	    }
	    
	XtPopdown(shells[winnum-1]);
	popped[winnum-1] = 0;
	currentview_xid = XtWindow(canvas);
	canvas_resize_proc(canvas, &event, &params, &nparams);
      
	while(XtAppPending(app_con))
	  {
	    XtAppNextEvent(app_con, &event);
	    XtDispatchEvent(&event);
	  }
      }
    else {
	printf("<bad arguments to deletewin, %s>\n",title) ;
    }
}
